logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/EphemeralMessageDeletionHandler.kt

```diff
@@ -0,0 +1,158 @@
+                coroutineScope = userSessionCoroutineScope
+            )
+
+            for (timerEvent in selfDeletingMessageTimer.startTimer(message.timeLeftForDeletion())) {
+                when (timerEvent) {
+                    is SelfDeletionTimerState.Started -> {
+                        if (!message.isDeletionScheduledInThePast()) {
+                            messageRepository.markSelfDeletionStartDate(
+                                message.conversationId,
+                                message.id,
+                                Clock.System.now().toEpochMilliseconds()
+                            )
+                        }
+                    }
+
+                    is SelfDeletionTimerState.OnGoing -> {
+                        println("time left ${timerEvent.timeLeft.timeLeftInMs}")
+                        ephemeralMessageDeletionTimeLeftMapMutex.withLock {
+                            ephemeralMessageDeletionTimeLeftMap.update { currentState ->
+                                currentState + ((message.conversationId to message.id) to timerEvent.timeLeft)
+                            }
+                        }
+                    }
+
+                    is SelfDeletionTimerState.Finished -> {
+                        messageRepository.deleteMessage(message.id, message.conversationId)
+                        ephemeralMessageDeletionTimeLeftMapMutex.withLock {
+                            ephemeralMessageDeletionTimeLeftMap.update { currentState ->
+                                currentState - (message.conversationId to message.id)
+                            }
+                        }
+                    }
+
+                }
+            }
+        }
+    }
+
+    override fun observePendingMessageDeletionState() = ephemeralMessageDeletionTimeLeftMap
+
+    override fun enqueuePendingSelfDeletionMessages() {
+        launch {
+            messageRepository.getEphemeralMessages()
+                .onSuccess { ephemeralMessages ->
+                    ephemeralMessages.forEach { ephemeralMessage ->
+                        require(ephemeralMessage is Message.Ephemeral)
+
+                        enqueueMessageDeletion(message = ephemeralMessage)
+                    }
+                }
+        }
+    }
+
+}
+
+internal class SelfDeletingMessageTimer(
+    private val coroutineScope: CoroutineScope
+) : CoroutineScope by coroutineScope {
+    private companion object {
+        const val TIMER_UPDATE_INTERVAL_IN_MILLIS = 1000L
+    }
+
+    fun startTimer(expireAfterMillis: Long) = produce {
+        send(SelfDeletionTimerState.Started(Clock.System.now().toEpochMilliseconds()))
+
+        var elapsedTime = 0L
+
+        while (elapsedTime < expireAfterMillis) {
+            delay(TIMER_UPDATE_INTERVAL_IN_MILLIS)
+
+            send(SelfDeletionTimerState.OnGoing(SelfDeletionTimeLeft(expireAfterMillis - elapsedTime)))
+
+            elapsedTime += TIMER_UPDATE_INTERVAL_IN_MILLIS
```

Comment:
Shouldn't we add the time that really passed between iterations? Every time it actually waits for 1s and executes a send so it can take more than 1s and after some time it can differ 😄

Reply:
all this is doing on consumer side is updating a value, if there is a time difference i think it would be some miliseconds, however with mutexes it could take longer indeed, good point I am going to take a look at it with this in mind.